### ISSUE TITLE
Supporting VictoriaMetric's Extended PromQL

### ISSUE THREAD

**Issue by AeroNotix** [jacksontj/promxy#253]
There are some nice convenience functions here: https://github.com/VictoriaMetrics/VictoriaMetrics/wiki/ExtendedPromQL which make writing queries, alerts etc much simpler.

When I try to use them with promxy, I get errors such as:

`Error executing query: invalid parameter 'query': parse error at char 3: unknown function with name "ru"` for a query such as: `ru(kubelet_volume_stats_available_bytes, kubelet_volume_stats_capacity_bytes)`. 

I understand you are using the underlying prometheus libraries for a lot of the query layer and it's likely that layer which is bubbling this error - but would promxy consider adding support for VM's extended PromQL?

**Comment by jacksontj** [comment]
I'd definitely consider it, I actually have an issue upstream (https://github.com/VictoriaMetrics/VictoriaMetrics/issues/56) to move the extended promql stuff into a more re-usable library. At this point I don't have the driving use case to spend my time rearchitecting that library to work with promql's types, but I'm definitely for it being in there eventually, PRs are definitely welcome!

**Comment by AeroNotix** [comment]
Cool, worthwhile changing the labels to feature or similar?

**Comment by jacksontj** [comment]
Done :)

**Comment by valyala** [comment]
FYI, we already have a PR that moves extended PromQL code into a separate library - https://github.com/VictoriaMetrics/VictoriaMetrics/pull/256

**Comment by valyala** [comment]
FYI, PromQL parser from VictoriaMetrics has been extracted into a standalone library - see [docs](https://godoc.org/github.com/VictoriaMetrics/VictoriaMetrics/lib/metricsql).

**Comment by jacksontj** [comment]
From a quick look it'll require some more work to be supported *in* promxy, as promxy actually uses a forked version of promql with support for `NodeReplacer` (a mechanism to rewrite the query tree post-parse pre-execution). I have patches to add that functionality on prometheus that I maintain, presumably we could add something similar here.

**Comment by valyala** [comment]
This sounds good!

**Comment by valyala** [comment]
FYI, the MetricsQL parser has been moved to a separate repository with its own release schedule - https://github.com/VictoriaMetrics/metricsql . This should significantly simplify its' integration into third-party projects such as Promxy.

**Comment by jeromegn** [comment]
I'd be very interested in this.

@jacksontj if you can point me at the right bits of code, I'm sure i could come up with a PR.

My company would be happy to sponsor this issue if you'd prefer that.

**Comment by jacksontj** [comment]
The place that I _think_ we'd want to implement this is to implement the Engine interface (https://github.com/jacksontj/promxy/blob/master/cmd/promxy/main.go#L188) and then add an option to swap between those.

TBH I haven't spent a lot of time looking into it so I'm not sure exactly how difficult it'll be. I'd be more than happy to review a PR and if you are interested in other options we can discuss those as well :)

**Comment by sdryga** [comment]
I use Promxy in congestion with VictoriaMetrics. I really lack the support of MetricsQL in Promxy. I'll be glad to see this in the next release. Hope it'll be implemented soon! )

**Comment by davidsome** [comment]
@jacksontj I use Promxy in conjunction with VictoriaMetrics. I really lack the support of MetricsQL in Promxy. I'll be glad to see this in the next release. Hope it'll be implemented soon! )

**Comment by dnull88** [comment]
+1 for supporting of MetricsQL in Promxy

**Comment by IrekFasikhov** [comment]
+1 for supporting of MetricsQL in promxy

**Comment by IrekFasikhov** [comment]
@jacksontj Hi, Thomas. Need some help for the implementation?

**Comment by ihard** [comment]
+1 very demanded functionality

**Comment by powershot275** [comment]
+1 Would love to see this implemented. Only thing stopping me from switching to promxy. 🙂

**Comment by anvpetrov** [comment]
+1 for supporting of MetricsQL in promxy

**Comment by autokilla47** [comment]
+1 for supporting of MetricsQL in promxy

**Comment by valdis932** [comment]
+

**Comment by AeroNotix** [comment]
yikes, lads, +1 city in here.

Pull out the editor and get to work. Sheesh.

**Comment by AeroNotix** [comment]
https://github.com/jacksontj/promxy/issues/253#issuecomment-569083961 @jacksontj where are these patches?

**Comment by G-Asura** [comment]
+1 for supporting of MetricsQL in promxy

**Comment by tiagosanto737** [comment]
+1 for supporting MetricsQL in promxy. This is extremely important for us to be able to use promxy in our production environment that has multiple victoriametrics clusters. We have 1000s of dashboards built using promQL and metricsQL and can not change them.

**Comment by hagen1778** [comment]
Does anyone work on this? If no, I'd like to take it)

**Comment by V1pr** [comment]
> Does anyone work on this? If no, I'd like to take it)

@valyala were you able to make any progress? :)

**Comment by oOHenry** [comment]
> I understand VM doesn't support remote read. Any workaround?
set `remote_read: false` in promxy config ;)

**Comment by mohammadkhavari** [comment]
Hey, Is it still in progress? @valyala @jacksontj

**Comment by sergeyshevch** [comment]
Hi! We're very interested in this feature and I will try to make PR within the next few days. @jacksontj Can you please guide me to the right entry point for such modifications?

Is this still relevant? (3 years after) https://github.com/jacksontj/promxy/issues/253#issuecomment-662597713

**Comment by sergeyshevch** [comment]
After codebase research, I've found the following points.

- We need to update the downstream Prometheus code to accept the `QueryEngine` interface in the `WebOptions` struct and in internal packages. (I will create PR for this)
- We need to implement QueryEngine for VictoriaMetrics. As I see we mostly need to bypass some checks of promql.Parser and use MetricsQL validation instead.
- The MetricsQL package has huge differences from the PromQL parser so it's hard to implement the query engine

It will be great if VictoriaMetrics has some QueryEngine implementation on their own like the Thanos project.
https://github.com/thanos-io/promql-engine

**Comment by isodude** [comment]
Does it not make sense to just support MetricsQL out of the box, but only if downstream targets support it?
I.e. adding a `QueryLanguage` variable to the ServerGroup model. If the query is to be sent to that one, MetricsQL is used to parse.

**Comment by jacksontj** [comment]
@isodude to be clear here; it would be nice to add support, but as @sergeyshevch mentioned there are a few issues with doing so.

In addition to the list @sergeyshevch mentioned (specifically adding support for a different engine in [Options](https://pkg.go.dev/github.com/prometheus/prometheus/web#Options)). We'd also need to add support for a [`NodeReplacer`](https://github.com/jacksontj/promxy/blob/master/pkg/proxystorage/proxy.go#L313) type interface -- as this is the main "secret sauce" that makes promxy performance what it is. This is what requires me to maintain a fork of prometheus (as it doesn't support this either).

**Comment by dvnscr** [comment]
Is there any hope for this to happen ?

**Comment by isodude** [comment]
@dvnscr I implemented with VictoriaMetrics cluster setup instead and the performance is much better since it's using a binary protocol after vmselect. I don't know if the effort of coaxing support into this project is worth it. There will be a lot of considerations to be made, since it differs in a lot of small details and is moving a bit faster.

It's actually breaking API-wise for /query by allowing '&step=' as a parameter.

**Comment by dvnscr** [comment]
@isodude Yeah, thanks, i am thinking about switching Single Victoria variant to Cluster stack on the single node per region for now. It will become future proof if i need to scale it horizontaly too 👍🏻
